### PR TITLE
Complement graph generator

### PR DESCRIPTION
**Pull request by jkinable**
Added a generator which generates the complement graph of a given input graph

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/generate/ComplementGraphGenerator.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    @Override
+    public void generateGraph(Graph<V, E> target, VertexFactory<V> vertexFactory, Map<String, V> resultMap) {
+        Graphs.addAllVertices(target, graph.vertexSet());
+
+        if (graph.getType().isDirected()) {
+            for (V u : graph.vertexSet())
+                for (V v : graph.vertexSet())
+                    if (u == v)
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/generate/ComplementGraphGenerator.java]: Here better use `equals()`.

Comment (jkinable) [reply jgrapht-core/src/main/java/org/jgrapht/generate/ComplementGraphGenerator.java]: Is equals() really necessary here? I iterate twice over the same set of vertices. The set is not modified in between. In this case, it seems impossible to have a case where u==v fails, but u.equals(v) is satisfied. The inner loop should only be skipped if u and v are physically the same object.

A simple "==" comparison is generally more efficient than an equals comparison.

Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/generate/ComplementGraphGenerator.java]: Right!

**Comment by d-michail** [comment]
Thanks

**Pull request merged**